File: Backend/app/routers/notes.py
from fastapi import APIRouter, Depends, status, Query
from typing import List, Optional
from supabase import Client
from app.dependencies import get_current_user, get_user_db_client
from app.schemas.user import UserResponse
from app.schemas.note import NoteCreate, NoteUpdate, NoteResponse
from app.services import note_service


router = APIRouter(prefix="/api/notes", tags=["notes"])


@router.get("", response_model=List[NoteResponse])
async def list_notes(
    search: Optional[str] = Query(None),
    favorites: Optional[bool] = Query(None),
    current_user: UserResponse = Depends(get_current_user),
    supabase: Client = Depends(get_user_db_client)
):
    """List all notes for the current user with optional filters."""
    return await note_service.get_notes(supabase, current_user.id, search, favorites)


@router.get("/trash", response_model=List[NoteResponse])
async def list_trash_notes(
    current_user: UserResponse = Depends(get_current_user),
    supabase: Client = Depends(get_user_db_client)
):
    """List all deleted notes for the current user."""
    return await note_service.get_trash_notes(supabase, current_user.id)


@router.get("/{note_id}", response_model=NoteResponse)
async def get_note(
    note_id: str,
    current_user: UserResponse = Depends(get_current_user),
    supabase: Client = Depends(get_user_db_client)
):
    """Get a single note by ID."""
    return await note_service.get_note_by_id(supabase, note_id, current_user.id)



@router.post("", response_model=NoteResponse, status_code=status.HTTP_201_CREATED)
async def create_note(
    note_data: NoteCreate,
    current_user: UserResponse = Depends(get_current_user),
    supabase: Client = Depends(get_user_db_client)
):
    """Create a new note."""
    return await note_service.create_note(supabase, current_user.id, note_data)


@router.put("/{note_id}", response_model=NoteResponse)
async def update_note(
    note_id: str,
    note_data: NoteUpdate,
    current_user: UserResponse = Depends(get_current_user),
    supabase: Client = Depends(get_user_db_client)
):
    """Update an existing note."""
    return await note_service.update_note(supabase, note_id, current_user.id, note_data)


@router.delete("/{note_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_note(
    note_id: str,
    current_user: UserResponse = Depends(get_current_user),
    supabase: Client = Depends(get_user_db_client)
):
    """Soft delete a note (move to trash)."""
    await note_service.delete_note(supabase, note_id, current_user.id)
    return None


@router.post("/{note_id}/restore", response_model=NoteResponse)
async def restore_note(
    note_id: str,
    current_user: UserResponse = Depends(get_current_user),
    supabase: Client = Depends(get_user_db_client)
):
    """Restore a note from trash."""
    return await note_service.restore_note(supabase, note_id, current_user.id)


@router.delete("/{note_id}/permanent", status_code=status.HTTP_204_NO_CONTENT)
async def permanent_delete_note(
    note_id: str,
    current_user: UserResponse = Depends(get_current_user),
    supabase: Client = Depends(get_user_db_client)
):
    """Permanently delete a note from database."""
    await note_service.permanent_delete_note(supabase, note_id, current_user.id)
    return None

File: Frontend/src/services/notes.js
import { apiClient } from './api'

/**
 * Notes service - handles all note-related API calls
 */
export const notesService = {
  /**
   * Get all notes (with optional filters)
   */
  async getNotes(filters = {}) {
    const params = new URLSearchParams()
    
    if (filters.search) {
      params.append('search', filters.search)
    }
    
    if (filters.favorites) {
      params.append('favorites', 'true')
    }
    
    const query = params.toString()
    const endpoint = query ? `/api/notes?${query}` : '/api/notes'
    
    return apiClient.get(endpoint)
  },

  /**
   * Get a single note by ID
   */
  async getNote(noteId) {
    return apiClient.get(`/api/notes/${noteId}`)
  },

  /**
   * Create a new note
   */
  async createNote(noteData) {
    return apiClient.post('/api/notes', {
      title: noteData.title || '',
      content: noteData.content || '',
      is_favorite: noteData.favorite || false,
    })
  },

  /**
   * Update an existing note
   */
  async updateNote(noteId, noteData) {
    return apiClient.put(`/api/notes/${noteId}`, {
      title: noteData.title,
      content: noteData.content,
      is_favorite: noteData.favorite,
    })
  },

  /**
   * Delete a note (soft delete - moves to trash)
   */
  async deleteNote(noteId) {
    return apiClient.delete(`/api/notes/${noteId}`)
  },

  /**
   * Get trash notes
   */
  async getTrashNotes() {
    return apiClient.get('/api/notes/trash')
  },

  /**
   * Restore a note from trash
   */
  async restoreNote(noteId) {
    return apiClient.post(`/api/notes/${noteId}/restore`, {})
  },

  /**
   * Permanently delete a note
   */
  async permanentDeleteNote(noteId) {
    return apiClient.delete(`/api/notes/${noteId}/permanent`)
  },

  /**
   * Generate a share link for a note
   */
  async shareNote(noteId) {
    return apiClient.post(`/api/notes/${noteId}/share`, {})
  },

  /**
   * Revoke share link for a note
   */
  async unshareNote(noteId) {
    return apiClient.delete(`/api/notes/${noteId}/share`)
  },

  /**
   * Get a shared note by token (public, no auth required)
   */
  async getSharedNote(token) {
    const API_BASE_URL = import.meta.env.VITE_API_BASE_URL || 'http://localhost:8000'
    const response = await fetch(`${API_BASE_URL}/api/share/${token}`)
    
    if (!response.ok) {
      throw new Error('Failed to fetch shared note')
    }
    
    return response.json()
  },
}
